# Netflix User Analytics
**Scenario:** Netflix wants to better understand user behaviour and use Machine Learning to improve customer retention and business decisions.

Dataset: 750 users with demographics, viewing habits, subscription details, and spending patterns.

In [1]:
import pandas as pd
import numpy as np
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.preprocessing import LabelEncoder

## Part A: Dataset Understanding

In [2]:
# Q1. Load the dataset and display the first five records
df = pd.read_csv('Netflix_Customers.csv')
df.head()

UserID,Age,Gender,SubscriptionType,WatchHoursPerWeek,DevicesUsed,FavoriteGenre,AdClicks,MonthlySpend,SubscriptionRenewed
1001,22,Female,Basic,7,1,Comedy,13,353,No
1002,55,Male,Premium,16,3,Drama,14,317,Yes
1003,49,Male,Basic,9,2,Sci-Fi,16,309,No
1004,39,Female,VIP,28,4,Horror,45,833,Yes
1005,38,Female,VIP,22,4,Romance,24,804,Yes


In [3]:
# Q2. Number of rows and columns
print('Shape:', df.shape)
print(f'Rows: {df.shape[0]}, Columns: {df.shape[1]}')

Shape: (750, 10)
Rows: 750, Columns: 10


In [4]:
# Q3. Display all column names
print('Column Names:')
print(df.columns.tolist())

Column Names:
['UserID', 'Age', 'Gender', 'SubscriptionType', 'WatchHoursPerWeek', 'DevicesUsed', 'FavoriteGenre', 'AdClicks', 'MonthlySpend', 'SubscriptionRenewed']


In [5]:
# Q4. Identify numerical and categorical features
print('Data Types:')
print(df.dtypes)

numerical = df.select_dtypes(include=['int64', 'float64']).columns.tolist()
categorical = df.select_dtypes(include=['object']).columns.tolist()

print('\nNumerical Features:', numerical)
print('Categorical Features:', categorical)

Data Types:
UserID                  int64
Age                     int64
Gender                 object
SubscriptionType       object
WatchHoursPerWeek       int64
DevicesUsed             int64
FavoriteGenre          object
AdClicks                int64
MonthlySpend            int64
SubscriptionRenewed    object
dtype: object

Numerical Features: ['UserID', 'Age', 'WatchHoursPerWeek', 'DevicesUsed', 'AdClicks', 'MonthlySpend']
Categorical Features: ['Gender', 'SubscriptionType', 'FavoriteGenre', 'SubscriptionRenewed']


In [6]:
# Q5. Check for missing values
print('Missing Values per Column:')
print(df.isnull().sum())
print('\nTotal Missing Values:', df.isnull().sum().sum())

Missing Values per Column:
UserID                 0
Age                    0
Gender                 0
SubscriptionType       0
WatchHoursPerWeek      0
DevicesUsed            0
FavoriteGenre          0
AdClicks               0
MonthlySpend           0
SubscriptionRenewed    0
dtype: int64

Total Missing Values: 0


## Part B: Exploratory Data Analysis

In [7]:
# Q6. Average age of users
avg_age = round(df['Age'].mean(), 2)
print('Average Age of Users:', avg_age)

Average Age of Users: 41.82


In [8]:
# Q7. Average watch hours per week
avg_watch = round(df['WatchHoursPerWeek'].mean(), 2)
print('Average Watch Hours Per Week:', avg_watch)

Average Watch Hours Per Week: 14.24


In [9]:
# Q8. Average monthly spending
avg_spend = round(df['MonthlySpend'].mean(), 2)
print('Average Monthly Spending (₹):', avg_spend)

Average Monthly Spending (₹): 689.91


In [10]:
# Q9. Number of users in each subscription category
print('Users per Subscription Type:')
print(df['SubscriptionType'].value_counts())

Users per Subscription Type:
SubscriptionType
Basic      342
Premium    279
VIP        129
Name: count, dtype: int64


In [11]:
# Q10. Percentage of users who renewed their subscription
renewed = df['SubscriptionRenewed'].value_counts(normalize=True) * 100
print('Subscription Renewal Percentage:')
print(renewed.round(2).astype(str) + '%')
print(f"\nUsers who Renewed: {renewed.get('Yes', 0):.2f}%")

Subscription Renewal Percentage:
SubscriptionRenewed
No     53.73%
Yes    46.27%

Users who Renewed: 46.27%


## Part C: Data Preparation

In [12]:
# Q11. Convert categorical features into numerical form using LabelEncoder
le = LabelEncoder()
df['Gender'] = le.fit_transform(df['Gender'])
df['SubscriptionType'] = le.fit_transform(df['SubscriptionType'])
df['FavoriteGenre'] = le.fit_transform(df['FavoriteGenre'])
df['SubscriptionRenewed'] = le.fit_transform(df['SubscriptionRenewed'])

print('After encoding:')
df.head()

After encoding:


   UserID  Age  Gender  SubscriptionType  WatchHoursPerWeek  DevicesUsed  FavoriteGenre  AdClicks  MonthlySpend  SubscriptionRenewed
0    1001   22       0                 0                  7            1              1        13           353                    0
1    1002   55       1                 1                 16            3              2        14           317                    1
2    1003   49       1                 0                  9            2              4        16           309                    0
3    1004   39       0                 2                 28            4              3        45           833                    1
4    1005   38       0                 2                 22            4              5        24           804                    1

In [13]:
# Q12. Define feature set (X) and target variable (y) for subscription renewal prediction
X = df.drop(['UserID', 'SubscriptionRenewed', 'MonthlySpend'], axis=1)
y = df['SubscriptionRenewed']

print('Features (X):', X.columns.tolist())
print('Target (y): SubscriptionRenewed')

Features (X): ['Age', 'Gender', 'SubscriptionType', 'WatchHoursPerWeek', 'DevicesUsed', 'FavoriteGenre', 'AdClicks']
Target (y): SubscriptionRenewed


In [14]:
# Q13. Split dataset into training and testing sets (80/20 split)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print('Training set size:', X_train.shape[0])
print('Testing set size: ', X_test.shape[0])

Training set size: 600
Testing set size:  150


## Part D: Decision Tree Classification

In [15]:
# Q14. Train a Decision Tree model to predict subscription renewal
dt = DecisionTreeClassifier(random_state=42)
dt.fit(X_train, y_train)
dt_pred = dt.predict(X_test)
print('Decision Tree model trained successfully.')

Decision Tree model trained successfully.


In [16]:
# Q15. Evaluate model using accuracy
dt_accuracy = accuracy_score(y_test, dt_pred)
print('Decision Tree Accuracy:', round(dt_accuracy * 100, 2), '%')

Decision Tree Accuracy: 55.33 %


In [17]:
# Q16. Generate and interpret the confusion matrix
cm = confusion_matrix(y_test, dt_pred)
print('Confusion Matrix:')
print(cm)
print('\nInterpretation:')
print(f'True Negatives  (Did not renew, correctly predicted): {cm[0][0]}')
print(f'False Positives (Did not renew, predicted as renewed): {cm[0][1]}')
print(f'False Negatives (Renewed, predicted as not renewed):  {cm[1][0]}')
print(f'True Positives  (Renewed, correctly predicted):       {cm[1][1]}')

Confusion Matrix:
[[49 33]
 [34 34]]

Interpretation:
True Negatives  (Did not renew, correctly predicted): 49
False Positives (Did not renew, predicted as renewed): 33
False Negatives (Renewed, predicted as not renewed):  34
True Positives  (Renewed, correctly predicted):       34


## Part E: K-Nearest Neighbors (KNN)

In [18]:
# Q17. Train a KNN classifier with K = 5
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)
knn_pred = knn.predict(X_test)
print('KNN model trained successfully.')

KNN model trained successfully.


In [19]:
# Q18. Compare accuracy of KNN with Decision Tree
knn_accuracy = accuracy_score(y_test, knn_pred)
print('KNN Accuracy:          ', round(knn_accuracy * 100, 2), '%')
print('Decision Tree Accuracy:', round(dt_accuracy * 100, 2), '%')

if dt_accuracy > knn_accuracy:
    print('\nDecision Tree performed better for subscription renewal prediction.')
elif knn_accuracy > dt_accuracy:
    print('\nKNN performed better for subscription renewal prediction.')
else:
    print('\nBoth models performed equally.')

KNN Accuracy:           58.67 %
Decision Tree Accuracy: 55.33 %

KNN performed better for subscription renewal prediction.


## Part F: Linear Regression

In [20]:
# Q19. Train a Linear Regression model to predict monthly spending
X_reg = df.drop(['UserID', 'SubscriptionRenewed', 'MonthlySpend'], axis=1)
y_reg = df['MonthlySpend']

X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(
    X_reg, y_reg, test_size=0.2, random_state=42
)

lr = LinearRegression()
lr.fit(X_train_reg, y_train_reg)
print('Linear Regression model trained successfully.')

Linear Regression model trained successfully.


In [21]:
# Q20. Predict monthly spending for a new user and interpret the result
# New user: Age=30, Gender=Male(1), SubscriptionType=Premium(1),
# WatchHoursPerWeek=15, DevicesUsed=2, FavoriteGenre=Action(0), AdClicks=10
new_user = pd.DataFrame([[30, 1, 1, 15, 2, 0, 10]],
                        columns=X_reg.columns)
predicted_spend = lr.predict(new_user)
print('Predicted Monthly Spending for New User: ₹', round(predicted_spend[0], 2))
print('\nInterpretation: Based on the user profile (Age=30, Gender=Male, SubscriptionType=Premium,')
print('WatchHoursPerWeek=15, DevicesUsed=2, FavoriteGenre=Action, AdClicks=10),')
print(f'the model predicts a monthly spending of ₹{round(predicted_spend[0], 2)}.')
print('This is above the dataset average of ₹689.91, suggesting this user is a relatively high spender.')

Predicted Monthly Spending for New User: ₹ 835.8

Interpretation: Based on the user profile (Age=30, Gender=Male, SubscriptionType=Premium,
WatchHoursPerWeek=15, DevicesUsed=2, FavoriteGenre=Action, AdClicks=10),
the model predicts a monthly spending of ₹835.8.
This is above the dataset average of ₹689.91, suggesting this user is a relatively high spender.


## Business Reflection Questions

**1. Which factors appear to influence subscription renewal the most?**  
Features like `WatchHoursPerWeek`, `AdClicks`, `SubscriptionType`, and `MonthlySpend` are likely the strongest predictors. Users who watch more and engage more with ads tend to be more invested in the platform and more likely to renew. VIP/Premium subscribers also show higher renewal intent.

**2. Why is subscription renewal a classification problem?**  
Because the output (`SubscriptionRenewed`) is a discrete category — either **Yes** or **No**. The model assigns each user to one of these two classes, making it a binary classification problem.

**3. Why is monthly spending a regression problem?**  
Because `MonthlySpend` is a continuous numerical value (e.g., ₹353, ₹835.8). Regression predicts a value on a continuous scale rather than assigning a category, making it appropriate here.

**4. Which algorithm performed better for renewal prediction?**  
KNN (K=5) achieved **58.67%** accuracy vs Decision Tree at **55.33%**. KNN performed slightly better on this dataset. The relatively low accuracy for both suggests that renewal behaviour may depend on factors not fully captured in the current features, or that the data has high variance in renewal patterns.

**5. How could the platform use these predictions to improve customer retention?**  
- **Identify at-risk users** (predicted not to renew) and target them with personalised discount offers or loyalty rewards before their renewal date.  
- **Predict high-value users** using the Linear Regression model and offer them VIP upgrade incentives to increase spending.  
- **Personalise content** recommendations based on `FavoriteGenre` to boost engagement (`WatchHoursPerWeek`), which in turn increases renewal likelihood.  
- **Optimise ad strategy** by targeting users with historically low `AdClicks` with more relevant ads to improve engagement.